PrefectHQ/prefect
# Fix login to private Docker registries

Pulling images from private Docker registries fails because the `DockerContainer` and `DockerRegistry` blocks use separate copies of `DockerClient`. I observed the problem happening with private container registries on both Docker Hub and Azure Container Registry.

Our current code might have assumed that calling `login` on the `docker-py` client behaves identically to `docker login` and saves credentials after a successful login. However, looking through its code, it does not appear to do so. It will read locally-stored credentials, but if it does not find any, it just stores the credentials we supply in a dict (see [here](https://github.com/docker/docker-py/blob/be942f83902fbd02e05270c39b6917880939c165/docker/api/daemon.py#L151) and then [here](https://github.com/docker/docker-py/blob/be942f83902fbd02e05270c39b6917880939c165/docker/auth.py#L315) in the `docker-py` code to verify).

This pull request fixes the issue by modifying `DockerRegistry` to return an authenticated instance of `DockerClient`, allowing `DockerContainer` to pull images from private registries.

Closes #6643 (Confirmed working by @sti0 in https://github.com/PrefectHQ/prefect/issues/6643#issuecomment-1254983958)

### Example
Previously:
```
if self.image_registry:
    self.image_registry.login()
    return docker_client.images.pull(image, tag)
```
Now:
```
if self.image_registry:
    # If an image registry block was supplied, load an authenticated Docker
    # client from the block. Otherwise, use an unauthenticated client to
    # pull images from public registries.
    docker_client = self.image_registry.get_docker_client()
else:
    docker_client = self._get_client()
       
...

return docker_client.images.pull(image, tag)
```

The downside to this approach is that it changes the interface of `BaseDockerLogin` and may have downstream impact on any classes that implement it. Outside of `DockerRegistry`, I believe there is an as yet unmerged PR to add ECR registry support.

We could instead run the Docker CLI to login and persist the credentials to disk if login is successful. This may have been the original intention as calling docker is mentioned in a comment in the code. Before doing this, we should consider:
* It will break on systems where Docker Engine is installed but Docker CLI is not. 
* On systems without a secure credential store available, we would end up saving credentials in plaintext in `~/.docker/config.json 
* Keeping the credentials in-memory where they exist only for the lifetime of the signed-in `DockerClient` probably best meets what users would expect when they choose to trust Prefect (either Orion or Cloud) with secret credentials.

### Steps to QA
To test the changes in this PR, I recommend:
* Run `docker logout` on all private registries you intend to test against, because `docker-py` will use them if it finds them.
* Delete all local copies of the images you intend to try to pull from private registries. Also delete any images you re-tagged and pushed to a private repo (i.e. if you tagged `prefecthq/prefect:2-python3.7` as `myprivatedockherhubacct/prefect:2` and pushed it to a private repo for testing, you'll need to delete your local copies of both `prefecthq/prefect:2-python3.7` and `myprivatedockherhubacct/prefect:2`). This is because even when you send an explicit image pull request to the Docker API, it ultimately [will not pull anything](https://github.com/docker/docker-ce/blob/44a430f4c43e61c95d4e9e9fd6a0573fa113a119/components/engine/distribution/pull_v2.go#L573) if an image with the same digest already exists locally.
* Set up registry and infrastructure blocks containing the details needed to pull the images you need. _Note:_ there was a UI bug in 2.4.2 when you view or edit a `DockerContainer` block. Image registry may show up as on the edit screen 'None' even if you previously had a registry selected, so you will need to re-select it in the dropdown before saving, otherwise your container block will be saved without an image registry. #6953 should fix this, so you will only notice it if testing a client from this PR's branch against a 2.4.2 API/UI.

### Checklist


	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes a `fix`, `feature`, `enhancement`, `docs`, `collections`, or `maintenance` label categorizing the change.
